pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,240 @@
+				if errors.As(err, &userErr) {
+					logger.Info(fmt.Sprintf("failed to fetch clientId for SPC for listener %s due to user error: %s, sending warning event", listener.Name, userErr.userMessage))
+					g.events.Eventf(gwObj, corev1.EventTypeWarning, "InvalidInput", "invalid TLS configuration: %s", userErr.userMessage)
+					return ctrl.Result{}, nil
+				}
+				logger.Error(err, fmt.Sprintf("failed to fetch clientId for listener %s: %s", listener.Name, err.Error()))
+				return ctrl.Result{}, fmt.Errorf("fetching clientId for listener: %w", err)
+			}
+
+			// otherwise it's active + valid - build SPC
+			certUri := string(listener.TLS.Options[certUriTLSOption])
+
+			logger.Info("building spc for listener and upserting")
+			spcConf := SPCConfig{
+				ClientId:        clientId,
+				TenantId:        g.config.TenantID,
+				KeyvaultCertUri: certUri,
+				Name:            generateGwListenerCertName(gwObj.Name, listener.Name),
+			}
+			err = buildSPC(spc, spcConf)
+			if err != nil {
+				var userErr userError
+				if errors.As(err, &userErr) {
+					logger.Info("failed to build SecretProviderClass from user error: %q sending warning event", userErr.userMessage)
+					g.events.Eventf(gwObj, corev1.EventTypeWarning, "InvalidInput", "invalid TLS configuration: %s", userErr.userMessage)
+					return ctrl.Result{}, nil
+				}
+				logger.Error(err, fmt.Sprintf("building SPC for listener %s: %s", listener.Name, err.Error()))
+				return ctrl.Result{}, fmt.Errorf("building spc: %w", err)
+			}
+
+			logger.Info(fmt.Sprintf("reconciling SecretProviderClass %s for listener %s", spc.Name, listener.Name))
+			if err = util.Upsert(ctx, g.client, spc); err != nil {
+				fullErr := fmt.Errorf("failed to reconcile SecretProviderClass %s: %w", req.Name, err)
+				logger.Error(err, fullErr.Error())
+				g.events.Event(gwObj, corev1.EventTypeWarning, "FailedUpdateOrCreateSPC", fullErr.Error())
+				return ctrl.Result{}, fullErr
+			}
+
+			logger.Info(fmt.Sprintf("preemptively attaching secret reference for listener %s", listener.Name))
+			newCertRef := gatewayv1.SecretObjectReference{
+				Namespace: to.Ptr(gatewayv1.Namespace(req.Namespace)),
+				Group:     to.Ptr(gatewayv1.Group(corev1.GroupName)),
+				Kind:      to.Ptr(gatewayv1.Kind("Secret")),
+				Name:      gatewayv1.ObjectName(generateGwListenerCertName(gwObj.Name, listener.Name)),
+			}
+			gwObj.Spec.Listeners[index].TLS.CertificateRefs = []gatewayv1.SecretObjectReference{newCertRef}
+			continue
+		}
+		// we should delete the SPC if it exists
+		logger.Info(fmt.Sprintf("attempting to remove unused SPC %s", spc.Name))
+
+		deletionSpc := &secv1.SecretProviderClass{}
+		if err = g.client.Get(ctx, client.ObjectKeyFromObject(spc), deletionSpc); err != nil {
+			if client.IgnoreNotFound(err) != nil {
+				logger.Error(err, fmt.Sprintf("failed to fetch SPC for deletion %s", spc.Name))
+				return ctrl.Result{}, fmt.Errorf("fetching SPC for deletion: %w", err)
+			}
+			return ctrl.Result{}, nil
```

Comment:
In the event there are listeners that are listenerIsKvEnabled and !listenerIsKvEnabled, you will return here without `Upserting` the gateway object with updated certificate refs. This might be a `continue` instead of a return.

Reply:
great catch! fixed + added a UT for this